### ISSUE TITLE
Non-deterministic NER results 

### ISSUE THREAD

**Issue by isLouisHsu** [princeton-nlp/PURE#18]
Hi Zexuan,

I tried to reproduct the NER results for SciERC and ran the same command 5 times(the default seed is 0). 
``` sh
python run_entity.py \
    --do_train \
    --context_window 0 \
    --task scierc \
    --data_dir scierc_data/processed_data/json/ \
    --model allenai/scibert_scivocab_uncased \
    --output_dir scierc_models/ent-scib-ctx0/
```

**However, I got non-deterministic results.**
**I've only observed this happening on the GPU, rather than the CPU.** 
**Is it possible to avoid the non-deterministic results?**
``` sh
08/19/2021 21:45:10 - INFO - root - Evaluating...
08/19/2021 21:45:13 - INFO - root - Accuracy: 0.989835
08/19/2021 21:45:13 - INFO - root - Cor: 1113, Pred TOT: 1703, Gold TOT: 1685
08/19/2021 21:45:13 - INFO - root - P: 0.65355, R: 0.66053, F1: 0.65702
08/19/2021 21:45:13 - INFO - root - Used time: 3.658346
08/19/2021 21:45:17 - INFO - root - Total pred entities: 1703

08/19/2021 21:45:32 - INFO - root - Evaluating...
08/19/2021 21:45:36 - INFO - root - Accuracy: 0.989780
08/19/2021 21:45:36 - INFO - root - Cor: 1135, Pred TOT: 1759, Gold TOT: 1685
08/19/2021 21:45:36 - INFO - root - P: 0.64525, R: 0.67359, F1: 0.65912
08/19/2021 21:45:36 - INFO - root - Used time: 3.603827
08/19/2021 21:45:39 - INFO - root - Total pred entities: 1759

08/19/2021 21:45:54 - INFO - root - Evaluating...
08/19/2021 21:45:58 - INFO - root - Accuracy: 0.989813
08/19/2021 21:45:58 - INFO - root - Cor: 1147, Pred TOT: 1769, Gold TOT: 1685
08/19/2021 21:45:58 - INFO - root - P: 0.64839, R: 0.68071, F1: 0.66416
08/19/2021 21:45:58 - INFO - root - Used time: 3.655415
08/19/2021 21:46:01 - INFO - root - Total pred entities: 1769

08/19/2021 21:46:17 - INFO - root - Evaluating...
08/19/2021 21:46:21 - INFO - root - Accuracy: 0.990009
08/19/2021 21:46:21 - INFO - root - Cor: 1107, Pred TOT: 1672, Gold TOT: 1685
08/19/2021 21:46:21 - INFO - root - P: 0.66208, R: 0.65697, F1: 0.65952
08/19/2021 21:46:21 - INFO - root - Used time: 3.643370
08/19/2021 21:46:24 - INFO - root - Total pred entities: 1672

08/19/2021 21:46:39 - INFO - root - Evaluating...
08/19/2021 21:46:43 - INFO - root - Accuracy: 0.989835
08/19/2021 21:46:43 - INFO - root - Cor: 1115, Pred TOT: 1714, Gold TOT: 1685
08/19/2021 21:46:43 - INFO - root - P: 0.65053, R: 0.66172, F1: 0.65608
08/19/2021 21:46:43 - INFO - root - Used time: 3.624065
08/19/2021 21:46:46 - INFO - root - Total pred entities: 1714
```

The GPU is GTX3090(CUDA 11.1). The packages installed are
``` txt
allennlp==2.4.0
torch==1.7.0
transformers==4.5.1
overrides==3.1.0
requests==2.24.0
```

Best,
Yaobin

**Comment by a3616001** [comment]
Hi Yaobin! How many GPUs are you using for each run? I suspect that this is because the model behavior might be slightly different with different #GPUs. (My experiments were based on one single 2080 GPU).

**Comment by isLouisHsu** [comment]
Hi Zexuan,

I used a single GTX3090 GPU. I found the problem is caused by `allennlp.nn.utils.batched_index_select` which calls `torch.index_select`.  I implemented the function and got deterministic results. However, the function is really time-consuming.
``` py
def batched_index_select(target: torch.Tensor, indices: torch.LongTensor) -> torch.Tensor:
    batch_size, sequence_length, hidden_size = target.size()
    _, num_spans = indices.size()
    indexed = []
    for b in range(batch_size):
        target_b = target[b]
        indices_b = indices[b]
        indexed_b = []
        for i in range(num_spans):
            indexed_b.append(target_b[indices_b[i]])
        indexed_b = torch.stack(indexed_b, dim=0)
        indexed.append(indexed_b)
    indexed = torch.stack(indexed, dim=0)
    return indexed
```

Best,
Yaobin
